Create the cognoml package to implement an MVP API

dhimmel opened the pull request:
I'll remove WIP (work in progress) when this pull request is complete.

See https://github.com/cognoma/machine-learning/issues/31 for API design discussion.

See https://github.com/cognoma/machine-learning/issues/44 for MVP (minimal viable product) feature inclusion decisions.

dhimmel committed:
Begin constructing a MVP machine learner

dhimmel committed:
Export JSON API input for Hippo pathway

dhimmel committed:
Ignore __pycache__

dhimmel committed:
classify() functioning with mock input

dhimmel committed:
Save output corresponding to hippo-input.json

From the `cognoml` directory, ran:

```
python analysis.py > ../data/api/hippo-output.json
```

dhimmel committed:
Export model information to JSON output

Also filter zero-variance features.

dhimmel commented on cognoml/analysis.py:
```diff
@@ -85,6 +86,11 @@ def classify(sample_id, mutation_status, **kwargs):
+    
+    results['model'] = utils.model_info(clf_grid.best_estimator_)
+
+    feature_df = utils.get_feature_df(clf_grid.best_estimator_, X.columns)
```
@yl565 currently I'm retrieving feature names using `X.columns` which will get the feature names of `X` before it enters the pipeline. However, since `VarianceThreshold` or other feature selection/tranformation steps will alter the feature set, do you know how we can get feature names at the end of the pipeline? In other words, we want the feature names corresponding to `clf_grid.best_estimator_.coef_`. I searched for like an hour and couldn't figure this out.

dhimmel replied:
See https://github.com/scikit-learn/scikit-learn/issues/7536

dhimmel committed:
Return unselected observations

Unselected observations (samples in the dataset that were not selected
by the user) are now returned. These observations receive predictions
but are missing (-1 encoded) for fields such as `testing` and `status`.

Sorted model parameters by key.

dhimmel committed:
Save grid_search performance metrics

dhimmel commented on cognoml/analysis.py:
```diff
@@ -56,26 +56,34 @@ def classify(sample_id, mutation_status, **kwargs):
+        ('sample_id', X_whole.index),
+        ('predicted_status', pipeline.predict(X_whole)),
+        ('predicted_score', pipeline.decision_function(X_whole)),
+        ('predicted_prob', pipeline.predict_proba(X_whole)[:, 1]),
```
@yl565 what's the best way to see if a pipeline supports `predict_proba`? We can upgrade to sklearn 18 once that's released, if that will make things easier.

dhimmel replied:
See d52c6a2a60615951e03b0e84eadbe07d718e908e for my solution

dhimmel committed:
Move classifier and pipeline to it's own module

dhimmel committed:
Add setup.py to make module installable

dhimmel commented:
This pull request is ready for some review. Suggesting @yl565 @awm33 @cgreene @gwaygenomics.

The package name is `cognoml`. Feedback on name welcome.

The package is pip installable, but there are outstanding issues:
- **data management**: I'm not sure the best way to have our package access the data. Should you always have to pass a path for the directory that contains the data? If the data doesn't exist in that directory, it can be downloaded.
- **algorithm support**: currently only SGDClassifier is supported. It's not clear how we want to modularly support different pipelines. If we want to support different algos, how should this info get passed to this module?
- [**feature name bug**](https://github.com/cognoma/machine-learning/pull/51#discussion_r79705959)
- currently the returned JSON may contain numpy values which break the builtin JSON encoder. Alternatively, `cognoml.utils.JSONEncoder` should work. However, if writing to the database, maybe we should just sanitize all JSON values before returning.

To run the example, use:

``` sh
python cognoml/main.py > data/api/hippo-output.json
```

gwaybio commented on cognoml/analysis.py:
```diff
@@ -0,0 +1,103 @@
+
+    performance = collections.OrderedDict()
+    for part, df in ('training', obs_train_df), ('testing', obs_test_df):
+        y_true=df.status
```
pep8 spacing updates needed

gwaybio commented on scripts/3.TCGA-MLexample_Pathway.py:
```diff
@@ -111,17 +112,17 @@
 # In[9]:
 
-get_ipython().run_cell_magic('time', '', "path = os.path.join('data', 'expression-matrix.tsv.bz2')\nX = pd.read_table(path, index_col=0)")
+get_ipython().run_cell_magic('time', '', "path = os.path.join('download', 'expression-matrix.tsv.bz2')\nX = pd.read_table(path, index_col=0)")
```
if someone were to nbconvert the `ipynb` file will this script be overwritten?

dhimmel replied:
Shouldn't be as the upstream changes to `3.TCGA-MLexample_Pathway.ipynb` are part of this pull request.

gwaybio replied:
ah ok, see it now

gwaybio commented on cognoml/analysis.py:
```diff
@@ -0,0 +1,103 @@
+    X = X_whole.loc[obs_df.sample_id, :]
+    y = obs_df.status
+    
+    X_train, X_test, y_train, y_test = train_test_split(
```
Confirming that you're deciding not to stratify based on disease too?

dhimmel replied:
Ah, stratification by disease could also make sense. Currently, sample/covariate info is not part of this pull request. I think it probably should be added before the first release.

gwaybio replied:
also, I was at talk by Olivier Elemento - he was building models for a different purpose (predict immunotherapy responders) but was adjusting for mutation burden as a covariate. We may want to consider checking out his stuff and adjusting for burden too

gwaybio commented on cognoml/analysis.py:
```diff
@@ -0,0 +1,103 @@
+    results['model']['features'] = utils.df_to_datatables(feature_df)
+
+    results['observations'] = utils.df_to_datatables(obs_df)
+    return results
```
Maybe in the beginning of this script you can describe what `results` should look like? I am having some difficulty interpreting what results actually entails and its format

dhimmel replied:
I generated a [JSON schema](http://json-schema.org/) using ([`hippo-output-schema.json`](https://github.com/dhimmel/machine-learning/blob/28cb22b1f7ec6c3438241e617d1bbc6d94a1410c/data/api/hippo-output-schema.json)):

```
genson --indent=2 data/api/hippo-output.json  > data/api/hippo-output-schema.json
```

I can add descriptions for each field here. Do you think that's a good solution.

gwaybio replied:
Maybe just a link to the hippo-output-schema and the genson command would suffice

gwaybio reviewed (commented):
Nice work. I only made a couple minor comments. I can also confirm that the example `data/api/hippo-output.json` works as expected!

dhimmel committed:
Review comments: spacing and results doc

gwaybio commented:
ok, i think this looks good to me. Although you may want to wait on @yl565 for those specific questions.

dhimmel committed:
Check whether pipeline has function before calling

Meant to address https://git.io/vPvtI

dhimmel committed:
Acquire data from figshare

dhimmel committed:
Update for sklearn 0.18.0, Fix pipeline

Fix pipeline according to:
https://github.com/scikit-learn/scikit-learn/issues/7536#issuecomment-251236645

Extract selected feature names according to:
https://github.com/scikit-learn/scikit-learn/issues/7536#issuecomment-251403876

dhimmel committed:
Merge branch 'master' into package

Used `git checkout --theirs` to resolve conflicts

dhimmel committed:
Semantic improvements of get_feature_df

dhimmel committed:
Update API JSON files

dhimmel committed:
Mention hippo-output-schema.json in docstring

dhimmel commented:
I propose we merge this sooner rather than later. The pull request is already quite large.

@gwaygenomics, @cgreene, @yl565 does one of you want to submit a review for up to 2291a0cd0afc74d3888cf04a52a169ca3435de19?

gwaybio commented on README.md:
````diff
@@ -20,3 +20,13 @@ conda env update --file environment.yml
+pip install --editable .
+```
+
+Make sure the `cognoma-machine-learning` environment is activated first, so the `cognoml` is only installed for this environment.
````
Would it make sense to say this requirement before `pip install`?

gwaybio commented on cognoml/analysis.py:
```diff
@@ -0,0 +1,114 @@
+    """
+    Read data.
+    """
+    v_dir = download_files(directory=data_directory, artile_id=3487685, version=version)
```
typo in `article_id`

also, is there a reason you are not providing `article_id` and `directory` as function arguments?

dhimmel replied:
> also, is there a reason you are not providing article_id and directory as function arguments?

What do you mean?

gwaybio replied:
the `read_data()` function has only a single argument: `version` - you have `download_files` here accepting two variables `data_directory` and `3487685` that are blind to the function.

If this was the functionality you are intending, I would recommend doing this instead:

``` python
def read_data(directory=data_directory, article_id=3487685, version=None):
    """
    Read data.
    """
    v_dir = download_files(directory=directory, article_id=article_id, version=version)
    ...
```

dhimmel replied:
`data_directory` is a module scope variable that should be set using:

``` python
cognoml.analysis.data_directory = 'new_path'
```

Using the default definition of `directory=data_directory` will break the above functionality.

There is no support for changing `article_id` currently.

So while I see your point, I think the changes add more complexity without any additional functionality at the moment.

dhimmel replied:
@gwaygenomics [note that](http://docs.python-guide.org/en/latest/writing/gotchas/#mutable-default-arguments):

> Python’s default arguments are evaluated once when the function is defined, not each time the function is called (like it is in say, Ruby).

gwaybio replied:
For mutable defaults only?

dhimmel replied:
> For mutable defaults only?

No for all defaults. Their value is evaluated upon definition. This leads to an odd behavior for mutable defaults where they can be modified with each function call. Immutables don't have this issue, but are still evaluated upon definition.

gwaybio commented on cognoml/analysis.py:
```diff
@@ -0,0 +1,114 @@
+    y = obs_df.status
+    
+    X_train, X_test, y_train, y_test = train_test_split(
+        X, y, test_size=0.1, random_state=0, stratify=y)
```
again, this is always set to 10%?

dhimmel replied:
Yeah unless you have another heuristic that you think is better. Eventually we will probably need to smarten up here and potentially refuse to classify problems with less than a certain number of positives.

gwaybio commented on cognoml/classifiers/logistic_regression.py:
```diff
@@ -0,0 +1,30 @@
+    #'classify__alpha': 10.0 ** np.arange(-3, 2),
+    #'classify__l1_ratio': [0.0, 0.1, 0.2, 0.5, 0.8, 0.9, 1.0],
+    'classify__alpha': 10.0 ** np.arange(-1, 1),
+    'classify__l1_ratio': [0.0, 1.0],
```
Lasso or Ridge only?

dhimmel replied:
Fixing the pipeline (#54) really slowed things down because the tranformation steps are now performed separately for each CV fold. Therefore, I really cut down the grid. In retrospect, this grid is probably too small -- I'll increase it.

dhimmel replied:
See https://github.com/cognoma/machine-learning/issues/56 and 10308e0f40278409a257b4c6119359a63ca22a98

gwaybio commented on cognoml/figshare.py:
```diff
@@ -0,0 +1,60 @@
+    version_to_url = {d['version']: d['url'] for d in response.json()}
+    return version_to_url
+
+def download_files(directory, artile_id=3487685, version=None):
```
another typo in `article_id`

gwaybio commented on cognoml/figshare.py:
```diff
@@ -0,0 +1,60 @@
+    str
+        The version-specific DOI corresponding to the downloaded data.
+    """
+    version_to_url = get_article_versions(artile_id)
```
article_id

gwaybio commented on cognoml/figshare.py:
```diff
@@ -0,0 +1,60 @@
+
+def download_files(directory, artile_id=3487685, version=None):
+    """
+    Download files for a specific figshare article_id and version to the specified directory.
```
a usage note that it won't _really_ download to the specified dictionary - it will append the version too

gwaybio commented on cognoml/utils.py:
```diff
@@ -0,0 +1,132 @@
+    `json.dump` function with `cls=JSONEncoder`.
+    """
+    obj_str = pd.json.dumps(obj)
+    print(obj_str)
```
making sure you want to print this here

dhimmel replied:
Yeah this main function is currently only for running a single example / test case. Actual users will call the `cognoml.analysis.classify()` function.

gwaybio commented on data/api/hippo-output-schema.json:
```diff
@@ -0,0 +1,323 @@
+{
```
I didn't read this file

gwaybio reviewed (changes requested):
nice comments and flow - easy to review. I didn't try running this iteration yet (I believe I tested one previously). But I can run if you're not pressed for time.

I only had some minor fixes and general questions.

dhimmel committed:
Address @gwaygenomics review comments

Does not address "Lasso or Ridge only?"

dhimmel committed:
Grid search: optimize alpha not l1_ratio

Do not optimize `l1_ratio`. Instead use the default of 0.15. Search a denser
grid for `alpha`. See https://github.com/cognoma/machine-learning/issues/56
